// flatMatchMaker/test/FlatMatchMakerTests.js
/**
/*
* Flat Match Maker
*
* Copyright 2014 Raising the Floor - International
*
* Licensed under the New BSD license. You may not use this file except in
* compliance with this License.
*
* The research leading to these results has received funding from the European Union's
* Seventh Framework Programme (FP7/2007-2013)
* under grant agreement no. 289016.
*
* You may obtain a copy of the License at
* https://github.com/GPII/universal/blob/master/LICENSE.txt
 */

/*global require, __dirname, module*/
(function () {
    "use strict";

    var fluid = require("infusion"),
        path = require("path"),
        jqUnit = fluid.require("jqUnit"),
        configPath = path.resolve(__dirname, "./configs"),
        kettle = fluid.registerNamespace("kettle"),
        gpii = fluid.registerNamespace("gpii"),
        fs = require("fs");

    fluid.require("flatMatchMaker", require);
    // fluid.require("kettle/test/utils/js/KettleTestUtils", require);


    fluid.registerNamespace("gpii.flatMatchMaker.tests");

    //gpii.flatMatchMaker.test.solutionsRegistry = require("../../../testData/win32.json")

    var fixtures = [
        {
            name: "No existing prefs in original set",
            preferences: {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {}
                    }
                }
            },
            inferred: {
                "gpii-default": {
                    "org.gnome.desktop.interface": {
                        "http://registry.gpii.net/common/fontSize": 9,
                        "http://registry.gpii.net/common/cursorSize": 0.9
                    },
                    "org.gnome.desktop.a11y.magnifier": {
                        "http://registry.gpii.net/common/magnification": 1.5,
                        "http://registry.gpii.net/common/magnifierPosition": "FullScreen"
                    }
                }
            },
            expect: {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/common/fontSize": 9,
                            "http://registry.gpii.net/common/cursorSize": 0.9,
                            "http://registry.gpii.net/common/magnification": 1.5,
                            "http://registry.gpii.net/common/magnifierPosition": "FullScreen"
                        }
                    }
                }
            }
        }, {
            name: "Existing prefs in original set, some being overwritten",
            preferences: {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/applications/org.gnome.desktop.interface": {
                                "cursor-size": 90,
                                "text-scaling-factor": 0.75
                            },
                            "http://registry.gpii.net/applications/org.alsa-project": {
                                "masterVolume": 50
                            },
                            "http://registry.gpii.net/common/cursorSize": 1
                        }
                    }
                }
            },
            inferred: {
                "gpii-default": {
                    "org.gnome.desktop.interface": {
                        "http://registry.gpii.net/common/fontSize": 9,
                        "http://registry.gpii.net/common/cursorSize": 0.9
                    },
                    "org.gnome.desktop.a11y.magnifier": {
                        "http://registry.gpii.net/common/magnification": 1.5,
                        "http://registry.gpii.net/common/magnifierPosition": "FullScreen"
                    }
                }
            },
            expect: {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/applications/org.gnome.desktop.interface": {
                                "cursor-size": 90,
                                "text-scaling-factor": 0.75
                            },
                            "http://registry.gpii.net/applications/org.alsa-project": {
                                "masterVolume": 50
                            },
                            "http://registry.gpii.net/common/fontSize": 9,
                            "http://registry.gpii.net/common/cursorSize": 1,
                            "http://registry.gpii.net/common/magnification": 1.5,
                            "http://registry.gpii.net/common/magnifierPosition": "FullScreen"
                        }
                    }
                }
            }
        }, {
            name: "Two instances of the same common term present in the inferred common terms",
            preferences: {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/common/cursorSize": 1
                        }
                    }
                }
            },
            inferred: {
                "gpii-default": {
                    "org.gnome.desktop.interface": {
                        "http://registry.gpii.net/common/fontSize": 9,
                        "http://registry.gpii.net/common/cursorSize": 0.9
                    },
                    "org.gnome.desktop.a11y.magnifier": {
                        "http://registry.gpii.net/common/fontSize": 1.5,
                        "http://registry.gpii.net/common/magnifierPosition": "FullScreen"
                    }
                }
            },
            expect: {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/common/fontSize": 9,
                            "http://registry.gpii.net/common/cursorSize": 1,
                            "http://registry.gpii.net/common/magnifierPosition": "FullScreen"
                        }
                    }
                }
            }
        }
    ];

    fluid.defaults("gpii.flatMatchMaker.tests", {
        gradeNames: ["fluid.test.testEnvironment", "autoInit"],
        components: {
            tester2: {
                type: "gpii.flatMatchMaker.tests.testCaseHolder"
            }
        }
    });

    fluid.defaults("gpii.flatMatchMaker.tests.testCaseHolder", {
        gradeNames: ["fluid.test.testCaseHolder", "autoInit"],
        // components: {
        //     "flatMatchMaker": {
        //         type: "gpii.flatMatchMaker"
        //     }
        // },
        modules: [{
            name: "flatMatchMakerTests",
            tests: [{
                expect: 3,
                name: "gpii.flatMatchMaker.addInferredCommon tests",
                func: "gpii.flatMatchMaker.tests.addInferredCommon"
            }]
        }]
    });

    gpii.flatMatchMaker.tests.addInferredCommon = function() {
        jqUnit.module("flatMatchMaker");
        jqUnit.test("gpii.flatMatchMaker.addInferredCommon tests", function() {
            fluid.each(fixtures, function (fixture) {
                var result = gpii.flatMatchMaker.addInferredCommon(fixture.preferences, fixture.inferred);
                jqUnit.assertDeepEq(fixture.name, fixture.expect, result);
            });
        });
    };

    // if (kettle.tests.allTests) {
    //     module.exports = "gpii.flatMatchMaker.tests";
    // } else {
        fluid.test.runTests(["gpii.flatMatchMaker.tests"]);
    // }
})();
