// flatMatchMaker/test/FlatMatchMakerTests.js
/*
* Flat Match Maker Tests
*
* Copyright 2014 Raising the Floor - International
*
* Licensed under the New BSD license. You may not use this file except in
* compliance with this License.
*
* The research leading to these results has received funding from the European Union's
* Seventh Framework Programme (FP7/2007-2013)
* under grant agreement no. 289016.
*
* You may obtain a copy of the License at
* https://github.com/GPII/universal/blob/master/LICENSE.txt
*/

/* eslint strict: ["error", "function"] */

// TODO: This should be converted to a set of browser tests

(function () {
    "use strict";

    var fluid = require("infusion"),
        jqUnit = fluid.registerNamespace("jqUnit"),
        kettle = require("kettle"),
        gpii = fluid.registerNamespace("gpii");

    kettle.loadTestingSupport();

    require("flatMatchMaker");
    require("ontologyHandler"); // TODO: We don't require to depend on the entire ontologyHandler
    require("matchMakerFramework");

    fluid.defaults("gpii.flatMatchMaker.tests", {
        gradeNames: ["fluid.test.testEnvironment"],
        components: {
            tester: {
                type: "gpii.flatMatchMaker.tests.testCaseHolder"
            }
        }
    });

    var disposeStrategyFixtures = [
        {
            description: "Basic test for disposeStrategy function",
            leavesArg: [
                "display.screenEnhancement.fontSize",
                "display.screenEnhancement.magnification",
                "display.screenEnhancement.-provisional-magnification/enabled",
                "display.screenReader.-provisional-screenReaderTTS/enabled"
            ],
            solRecsArg: {
                "test.dummy": {
                    "solution": {
                        "name": "Test dummy",
                        "contexts": {},
                        "settingsHandlers": {
                            "myconf": {
                                "type": "gpii.settingsHandlers.JSONSettingsHandler.set",
                                "options": {
                                    "path": "/some/path/to/file.json"
                                },
                                "capabilitiesTransformations": {
                                    "setting1": "http://registry\\.gpii\\.net/common/fontSize",
                                    "setting2.path1": "http://registry\\.gpii\\.net/common/screenReaderTTS/enabled",
                                    "setting2.path2": "http://registry\\.gpii\\.net/common/magnification/enabled",
                                    "setting2.path3.hello": "http://registry\\.gpii\\.net/common/magnification"
                                }
                            }
                        }
                    },
                    "skeleton": {
                        "applications": {
                            "test.dummy": {
                                "id": {}
                            }
                        },
                        "display": {
                            "screenEnhancement": {
                                "fontSize": {},
                                "magnification": {},
                                "-provisional-magnification/enabled": {}
                            },
                            "screenReader": {
                                "-provisional-screenReaderTTS/enabled": {}
                            }
                        }
                    },
                    "index": "test.dummy"
                },
                "test.dummy_2": {
                    "solution": {
                        "name": "Test dummy 2",
                        "contexts": {},
                        "settingsHandlers": {
                            "myconf": {
                                "type": "gpii.settingsHandlers.noSettings"
                            }
                        }
                    },
                    "skeleton": {
                        "applications": {
                            "test.dummy_2": {
                                "id": {}
                            }
                        }
                    },
                    "index": "test.dummy_2"
                }
            },
            expect: {
                "test.dummy_2": {
                    disposition: "reject"
                },
                "test.dummy": {
                    disposition: "accept"
                }
            }
        }
    ];

    gpii.flatMatchMaker.tests.disposeStrategy = function () {
        fluid.each(disposeStrategyFixtures, function (fixture) {
            var result = gpii.flatMatchMaker.disposeStrategy(fixture.leavesArg, fixture.solRecsArg);
            var expected = fluid.extend(true, {}, fixture.solRecsArg, fixture.expect);

            jqUnit.assertDeepEq(fixture.description, expected, result);
        });
    };

    var matchFixtures = [
        {
            description: "Full match call test",
            input: require("./data/pre_mm_payload1.json"),
            expect: {
                "inferredConfiguration": {
                    "gpii-default": {
                        "applications": {
                            "mac.dummy": {
                                "active": true,
                                "settings": {
                                    "http://registry.gpii.net/common/fontSize": 16,
                                    "http://registry.gpii.net/common/screenReaderTTS/enabled": false,
                                    "http://registry.gpii.net/common/magnification/enabled": false,
                                    "http://registry.gpii.net/common/magnification": 1
                                }
                            }
                        }
                    }
                }
            }
        }
    ];

    gpii.flatMatchMaker.tests.match = function (ontologyHandler) {
        fluid.each(matchFixtures, function (fixture) {
            fixture.input = gpii.matchMakerFramework.utils.preProcess(fixture.input);
            var result = gpii.flatMatchMaker.match(ontologyHandler, fixture.input, gpii.flatMatchMaker.disposeStrategy);
            jqUnit.assertDeepEq(fixture.description, fixture.expect, result);
        });
    };

    fluid.defaults("gpii.flatMatchMaker.tests.testCaseHolder", {
        gradeNames: ["fluid.test.testCaseHolder"],
        components: {
            "ontologyHandler": {
                type: "gpii.ontologyHandler"
            }
        },
        modules: [{
            name: "flatMatchMakerTests",
            tests: [{
                expect: 1,
                name: "gpii.flatMatchMaker.disposeStrategy tests",
                func: "gpii.flatMatchMaker.tests.disposeStrategy"
            }, {
                expect: 1,
                name: "gpii.flatMatchMaker.match tests",
                func: "gpii.flatMatchMaker.tests.match",
                args: "{ontologyHandler}"
            }]
        }]
    });

    module.exports = kettle.test.bootstrap("gpii.flatMatchMaker.tests");
})();
