// canopyMatchMaker/test/CanopyMatchMakerTests.js
/*
* Canopy Match Maker Tests
*
* Copyright 2015 Raising the Floor - International
*
* Licensed under the New BSD license. You may not use this file except in
* compliance with this License.
*
* The research leading to these results has received funding from the European Union's
* Seventh Framework Programme (FP7/2007-2013)
* under grant agreement no. 289016.
*
* You may obtain a copy of the License at
* https://github.com/GPII/universal/blob/master/LICENSE.txt
*/

/* global module */

// TODO: This should be converted to a set of browser tests

"use strict";

var fluid = require("infusion"),
    jqUnit = fluid.registerNamespace("jqUnit"),
    kettle = require("kettle"),
    gpii = fluid.registerNamespace("gpii");

kettle.loadTestingSupport();

require("canopyMatchMaker");
require("ontologyHandler"); // TODO: We don't require to depend on the entire ontologyHandler
require("matchMakerFramework");

fluid.registerNamespace("gpii.canopyMatchMaker.tests");

fluid.defaults("gpii.canopyMatchMaker.tests", {
    gradeNames: ["fluid.test.testEnvironment"],
    components: {
        tester: {
            type: "gpii.canopyMatchMaker.tests.testCaseHolder"
        }
    }
});

var matchFixtures = [
    {
        description: "Screenreader preference -> screenreader should be launched and configured",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/screenReaderTTS/enabled": true
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakescreenreader1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/screenReaderTTS/enabled": true
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "Screenreader preference (not matching apptology) -> screenreader should be launched and configured",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/pitch": 0.1
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakescreenreader1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/pitch": 0.1
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "Two magnifier preferences->best fit is launched, the other one should not run",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/invertColours": true
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/invertColours": true
                            }
                        },
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "Equally good magnifier fit->first hit is launched",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "magnifier and screen reader pref->both are launched",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/pitch": 0.8
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        },
                        "fakescreenreader1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/pitch": 0.8
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "APPLICATION PRIORITIES: prioritized magnifier is launched despite " +
            "the other one being first hit",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150
                },
                "metadata": [{
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag2"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        },
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "APPLICATION PRIORITIES: Different priority levels for magnifiers",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150
                },
                "metadata": [{
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag1"
                    ],
                    "value": 1024
                }, {
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag2"
                    ],
                    "value": 1025
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        },
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "APPLICATION PRIORITIES: Same priority levels for magnifiers," +
            " - where canopy uses first hit match",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150
                },
                "metadata": [{
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag1"
                    ],
                    "value": 1024
                }, {
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag2"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "APPLICATION PRIORITIES: Same priority levels for magnifiers" +
            " - best canopy fit selected",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/invertColours": true
                },
                "metadata": [{
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag1"
                    ],
                    "value": 1024
                }, {
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag2"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/invertColours": true
                            }
                        },
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "Application specific preference for magnifier - overrides canopy best fit",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/applications/fakemag2": {}
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/applications/fakemag2": {}
                            }
                        },
                        "fakemag1": {
                            "active": false,
                            "settings": {}
                        }
                    }
                }
            }
        }
    }, {
        description: "Application settings for multiple magnifiers - canopy best fit used",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/applications/fakemag2": {},
                    "http://registry.gpii.net/applications/fakemag1": {}
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/applications/fakemag1": {}
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/applications/fakemag2": {}
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "Application settings for app with no specific type",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/applications/otherApp": {}
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "otherApp": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/applications/otherApp": {}
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "Expl. user priorities overrides application settings implicit priority",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/applications/fakemag2": {},
                    "http://registry.gpii.net/common/magnification": 150
                },
                "metadata": [{
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag1"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/applications/fakemag2": {},
                                "http://registry.gpii.net/common/magnification": 150
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "/enabled support - when 'magnification/enabled: false' preference comes from transformation, we still have 'active: true'",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/magnification/enabled": false
                },
                "metadata": [{ // We select fakemag2
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag2"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": false
                            }
                        },
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": false
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "/enabled support - when 'magnification/enabled: true' preference comes from transformation, we still have 'active: true' only on relevant solution",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/magnification/enabled": true
                },
                "metadata": [{ // We select fakemag2
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag2"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": true
                            }
                        },
                        "fakemag2": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": true
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "/enabled support - when 'magnification/enabled: false' preference comes from capabilities block, we have 'active: false'",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/magnification/enabled": false
                },
                "metadata": [{ // We select fakemag1
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag1"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": false
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": false
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "/enabled support - when 'magnification/enabled: true' preference comes from capabilities block, we have 'active: true'",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/magnification": 150,
                    "http://registry.gpii.net/common/magnification/enabled": true
                },
                "metadata": [{ // We select fakemag1
                    "type": "priority",
                    "scope": [
                        "http://registry.gpii.net/applications/fakemag1"
                    ],
                    "value": 1024
                }]
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "fakemag1": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": true
                            }
                        },
                        "fakemag2": {
                            "active": false,
                            "settings": {
                                "http://registry.gpii.net/common/magnification": 150,
                                "http://registry.gpii.net/common/magnification/enabled": true
                            }
                        }
                    }
                }
            }
        }
    }, {
        description: "GPII-2074: Non-matching applications are ignored",
        input: require("./data/pre_mm_payload2.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/applications/notInstalledApp": {
                        "mypref": 20
                    }
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {}
                }
            }
        }
    }, {
        description: "GPII-3369: the top level high contrast theme is dispatched to both windows and UIO+",
        input: require("./data/pre_mm_payload3.json"),
        preferences: {
            "gpii-default": {
                "name": "Default preferences",
                "preferences": {
                    "http://registry.gpii.net/common/highContrast/enabled": true,
                    "http://registry.gpii.net/common/highContrastTheme": "white-black"
                }
            }
        },
        expect: {
            "inferredConfiguration": {
                "gpii-default": {
                    "applications": {
                        "com.microsoft.windows.highContrast": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/highContrast/enabled": true,
                                "http://registry.gpii.net/common/highContrastTheme": "white-black"
                            }
                        },
                        "net.gpii.uioPlus": {
                            "active": true,
                            "settings": {
                                "http://registry.gpii.net/common/highContrast/enabled": true,
                                "http://registry.gpii.net/common/highContrastTheme": "white-black"
                            }
                        }
                    }
                }
            }
        }
    }
];

gpii.canopyMatchMaker.tests.match = function (ontologyHandler) {
    fluid.each(matchFixtures, function (fixture) {
        fluid.set(fixture.input, ["preferences", "contexts"], fixture.preferences); // add testCase specific preferences
        var mmInput = gpii.matchMakerFramework.utils.preProcess(fixture.input); // run MM preprocessing to add datastructures
        var result = gpii.canopyMatchMaker.match(ontologyHandler, mmInput, gpii.canopyMatchMaker.utils.disposeStrategy);
        jqUnit.assertDeepEq(fixture.description, fixture.expect, result);
    });
};

fluid.defaults("gpii.canopyMatchMaker.tests.testCaseHolder", {
    gradeNames: ["fluid.test.testCaseHolder"],
    components: {
        "ontologyHandler": {
            type: "gpii.ontologyHandler"
        }
    },
    modules: [{
        name: "canopyMatchMakerTests",
        tests: [{
            expect: matchFixtures.length,
            name: "gpii.canopyMatchMaker.match tests",
            func: "gpii.canopyMatchMaker.tests.match",
            args: "{ontologyHandler}"
        }]
    }]
});

module.exports = kettle.test.bootstrap("gpii.canopyMatchMaker.tests");
