// flowManager/test/BrowserChannelTests.js
/**
 * GPII Flow Manager Browser Channel Tests
 *
 * Copyright 2013 OCAD University
 * Copyright 2015 Emergya
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

"use strict";

var fluid = require("infusion"),
    path = require("path"),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "./configs"),
    kettle = require("kettle"),
    gpii = fluid.registerNamespace("gpii");

fluid.require("ontologyHandler", require);
fluid.require("contextManager", require);
fluid.require("transformer", require);
fluid.require("lifecycleManager", require);
fluid.require("settingsHandlers", require);
fluid.require("matchMakerFramework", require);

kettle.loadTestingSupport();

fluid.registerNamespace("gpii.tests.flowManager.browserChannel");

gpii.tests.flowManager.browserChannel.chromeUserToken = "chrome2";
gpii.tests.flowManager.browserChannel.firefoxUserToken = "firefox";

gpii.tests.flowManager.browserChannel.testLoginResponseChrome = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.browserChannel.chromeUserToken + " was successfully logged in.", data);
};

gpii.tests.flowManager.browserChannel.testLogoutResponseChrome = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.browserChannel.chromeUserToken + " was successfully logged out.", data);
};

gpii.tests.flowManager.browserChannel.testLoginResponseFirefox = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.browserChannel.firefoxUserToken + " was successfully logged in.", data);
};

gpii.tests.flowManager.browserChannel.testLogoutResponseFirefox = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.browserChannel.firefoxUserToken + " was successfully logged out.", data);
};


gpii.tests.flowManager.browserChannel.checkConnectionRequest = function (data) {
    jqUnit.assertEquals("Connection suceeded", "A request to browserChannel has been successfully processed", data);
};

gpii.tests.flowManager.browserChannel.testDisconnection = function (data) {
    jqUnit.assertTrue("Disconnected", true);
};

gpii.tests.flowManager.browserChannel.checkWebSocketsComponent = function () {
    console.log("#### settings: " + JSON.stringify(gpii.settingsHandlers.WebSockets.getSettingsForId("org.chrome.cloud4chrome")));
    console.log("#### clients: " + Object.keys(gpii.settingsHandlers.WebSockets.model.clients));
};

gpii.tests.flowManager.browserChannel.checkClientOneNoSettings = function () {
    jqUnit.assertEquals("'org.chrome.cloud4chrome' is registered as a client",
        "org.chrome.cloud4chrome",
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients)
    );

    jqUnit.assertEquals("There aren't settings for 'org.chrome.cloud4chrome'",
        undefined,
        gpii.settingsHandlers.WebSockets.getSettingsForId("org.chrome.cloud4chrome")
    );
};

gpii.tests.flowManager.browserChannel.checkSettingsForChrome = function () {
    jqUnit.assertEquals("The settings for 'org.chrome.cloud4chrome' are the expected",
        JSON.stringify({"screenReaderTTSEnabled":false,"highContrastEnabled":true,"invertColours":false,"magnifierEnabled":true,"magnification":2,"fontSize":"medium","simplifier":false,"highContrastTheme":"white-black"}),
        JSON.stringify(gpii.settingsHandlers.WebSockets.getSettingsForId("org.chrome.cloud4chrome"))
    );
};

gpii.tests.flowManager.browserChannel.checkSettingsForFirefox = function () {
    jqUnit.assertEquals("The settings for 'org.mozilla.cloud4firefox' are the expected",
        JSON.stringify({"screenReaderTTSEnabled":false,"highContrastEnabled":true,"invertColours":false,"magnifierEnabled":true,"magnification":2,"fontSize":"medium","simplifier":false,"highContrastTheme":"black-white"}),
        JSON.stringify(gpii.settingsHandlers.WebSockets.getSettingsForId("org.mozilla.cloud4firefox"))
    );
};


gpii.tests.flowManager.browserChannel.checkNoClients = function () {
    jqUnit.assertEquals("No clients are registered yet", "", Object.keys(gpii.settingsHandlers.WebSockets.model.clients));
};

gpii.tests.flowManager.browserChannel.checkTwoChromeClients = function () {
    jqUnit.assertEquals("We have two clients",
        2,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients["org.chrome.cloud4chrome"]).length
    );
};

gpii.tests.flowManager.browserChannel.checkThreeClients = function () {
    jqUnit.assertEquals("We have two client ids",
        2,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients).length
    );

    jqUnit.assertEquals("We have two clients in 'org.chrome.cloud4chrome'",
        2,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients["org.chrome.cloud4chrome"]).length
    );

    jqUnit.assertEquals("We have one clients in 'org.mozilla.cloud4firefox'",
        1,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients["org.mozilla.cloud4firefox"]).length
    );
};

gpii.tests.flowManager.browserChannel.checkOneChromeOneFirefox = function () {
    jqUnit.assertEquals("We have two client ids",
        2,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients).length
    );

    jqUnit.assertEquals("We have one client in 'org.chrome.cloud4chrome'",
        1,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients["org.chrome.cloud4chrome"]).length
    );

    jqUnit.assertEquals("We have one client in 'org.mozilla.cloud4firefox'",
        1,
        Object.keys(gpii.settingsHandlers.WebSockets.model.clients["org.mozilla.cloud4firefox"]).length
    );
}

var testDefs = [{
    name: "Flow Manager BrowserChannel tests",
    //expect: 2,
    config: {
        configName: "browserChannel",
        configPath: configPath
    },
    components: {
        clientOne: {
            type: "kettle.test.request.io",
            options: {
                requestOptions: {
                    path: "/browserChannel",
                    port: "{configuration}.options.port"
                },
                listenOnInit: true
            }
        },
        clientTwo: {
            type: "kettle.test.request.io",
            options: {
                requestOptions: {
                    path: "/browserChannel",
                    port: "{configuration}.options.port"
                },
                listenOnInit: true
            }
        },
        clientThree: {
            type: "kettle.test.request.io",
            options: {
                requestOptions: {
                    path: "/browserChannel",
                    port: "{configuration}.options.port"
                },
                listenOnInit: true
            }
        },
        loginChrome: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    path: "/user/%userToken/login",
                    port: "{configuration}.options.port"
                },
                termMap: {
                    userToken: gpii.tests.flowManager.browserChannel.chromeUserToken
                }
            }
        },
        logoutChrome: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    path: "/user/%userToken/logout",
                    port: "{configuration}.options.port"
                },
                termMap: {
                    userToken: gpii.tests.flowManager.browserChannel.chromeUserToken
                }
            }
        },
        loginFirefox: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    path: "/user/%userToken/login",
                    port: "{configuration}.options.port"
                },
                termMap: {
                    userToken: gpii.tests.flowManager.browserChannel.firefoxUserToken
                }
            }
        },
        logoutFirefox: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    path: "/user/%userToken/logout",
                    port: "{configuration}.options.port"
                },
                termMap: {
                    userToken: gpii.tests.flowManager.browserChannel.firefoxUserToken
                }
            }
        }
    },
    sequence: [{
        func: "gpii.tests.flowManager.browserChannel.checkNoClients"
    }, {
        func: "{clientOne}.send",
        args: "org.chrome.cloud4chrome"
    }, {
        event: "{clientOne}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.checkConnectionRequest",
    }, {
/* Can I read what the endpoint is returning when doing the "connect" ???? */
/* Can I connect to gpii.webSockets.settingsHandlers.events.clientsModelChanged ??? */
        func: "gpii.tests.flowManager.browserChannel.checkClientOneNoSettings"
    }, {
        func: "{loginChrome}.send"
    }, {
        event: "{loginChrome}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLoginResponseChrome"
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkSettingsForChrome"
    }, {
        func: "{logoutChrome}.send"
    }, {
        event: "{logoutChrome}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLogoutResponseChrome"
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkClientOneNoSettings"
    }, {
        func: "{clientOne}.disconnect"
    }, {
        func: "{logoutChrome}.send"
    }, {
        event: "{logoutChrome}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLogoutResponseChrome"
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkNoClients"
/* clientOne's pass is done */
    }, {
        func: "{clientOne}.send",
        args: "org.chrome.cloud4chrome"
    }, {
        event: "{clientOne}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.checkConnectionRequest",
    }, {
        func: "{clientTwo}.send",
        args: "org.chrome.cloud4chrome"
    }, {
        event: "{clientTwo}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.checkConnectionRequest",
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkTwoChromeClients"
    }, {
        func: "{clientThree}.send",
        args: "org.mozilla.cloud4firefox"
    }, {
        event: "{clientThree}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.checkConnectionRequest",
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkThreeClients"
    }, {
        func: "{clientOne}.disconnect"
    }, {
        func: "{logoutChrome}.send"
    }, {
        event: "{logoutChrome}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLogoutResponseChrome"
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkOneChromeOneFirefox"
    }, {
        func: "{loginChrome}.send"
    }, {
        event: "{loginChrome}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLoginResponseChrome"
    }, {
        func: "{loginFirefox}.send"
    }, {
        event: "{loginFirefox}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLoginResponseFirefox"
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkSettingsForChrome"
    }, {
        func: "gpii.tests.flowManager.browserChannel.checkSettingsForFirefox"
    }, {
        func: "{logoutFirefox}.send"
    }, {
        event: "{logoutFirefox}.events.onComplete",
        listener: "gpii.tests.flowManager.browserChannel.testLogoutResponseFirefox"
/* To be continued ... */
    }]
}];

module.exports = kettle.test.bootstrapServer(testDefs);
