// flowManager/test/GetTokenTests.js
/**
 * GPII Flow Manager getToken Tests
 *
 * Copyright 2014 Raising the Floor - International
 * Copyright 2013 OCAD University
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

"use strict";

var fluid = require("infusion"),
    path = require("path"),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "./configs"),
    gpii = fluid.registerNamespace("gpii"),
    kettle = fluid.registerNamespace("kettle");

fluid.require("flowManager", require);

kettle.loadTestingSupport();

fluid.registerNamespace("gpii.tests.flowManager.getToken");

gpii.tests.flowManager.getToken.token = "testUser1";

gpii.tests.flowManager.getToken.testLoginResponse = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.getToken.token + " was successfully logged in.", data);
};

gpii.tests.flowManager.getToken.testLogoutResponse = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.getToken.token + " was successfully logged out.", data);
};

gpii.tests.flowManager.getToken.testInvalidGetTokenRequest = function (data) {
    data = JSON.parse(data);
    jqUnit.assertDeepEq("Expecting error messages as reply to http request", {
        isError: true,
        message: "No token(s) currently logged in to the system"
    }, data);
};

gpii.tests.flowManager.getToken.testGetTokenRequest = function (data) {
    data = JSON.parse(data);
    jqUnit.assertEquals("Expecting token returned in payload", data, gpii.tests.flowManager.getToken.token);
};

var testDefs = [{
    name: "Flow Manager getToken tests.",
    expect: 5,
    config: {
        configName: "getToken",
        configPath: configPath
    },
    components: {
        getTokenRequest: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    port: "{configuration}.options.port",
                    path: "/token"
                }
            }
        },
        login: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    port: "{configuration}.options.port",
                    path: "/user/%token/login"
                },
                termMap: {
                    token: gpii.tests.flowManager.getToken.token
                }
            }
        },
        logout: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    port: "{configuration}.options.port",
                    path: "/user/%token/logout"
                },
                termMap: {
                    token: gpii.tests.flowManager.getToken.token
                }
            }
        }
    },
    sequence: [{
        funcName: "fluid.pushSoftFailure",
        args: kettle.utils.failureHandler
    }, {
        func: "{getTokenRequest}.send"
    }, {
        event: "{getTokenRequest}.events.onComplete",
        listener: "gpii.tests.flowManager.getToken.testInvalidGetTokenRequest"
    }, {
        func: "{login}.send"
    }, {
        event: "{login}.events.onComplete",
        listener: "gpii.tests.flowManager.getToken.testLoginResponse"
    }, {
        func: "{getTokenRequest}.send"
    }, {
        event: "{getTokenRequest}.events.onComplete",
        listener: "gpii.tests.flowManager.getToken.testGetTokenRequest"
    }, {
        func: "{logout}.send"
    }, {
        event: "{logout}.events.onComplete",
        listener: "gpii.tests.flowManager.getToken.testLogoutResponse"
    }, {
        func: "{getTokenRequest}.send"
    }, {
        event: "{getTokenRequest}.events.onComplete",
        listener: "gpii.tests.flowManager.getToken.testInvalidGetTokenRequest"
    }, {
        funcName: "fluid.pushSoftFailure",
        args: -1
    }]
}];

module.exports = kettle.test.bootstrapServer(testDefs);
