// flowManager/test/GetUserTokenTests.js
/**
 * GPII Flow Manager getUserToken Tests
 *
 * Copyright 2014 Raising the Floor - International
 * Copyright 2013 OCAD University
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

"use strict";

var fluid = require("infusion"),
    path = require("path"),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "./configs"),
    gpii = fluid.registerNamespace("gpii"),
    kettle = fluid.require("kettle");

fluid.require("contextManager", require);
fluid.require("matchMakerFramework", require);
fluid.require("ontologyHandler", require);
fluid.require("transformer", require);

kettle.loadTestingSupport();

fluid.registerNamespace("gpii.tests.flowManager.getUserToken");

gpii.tests.flowManager.getUserToken.userToken = "testUser1";

gpii.tests.flowManager.getUserToken.testLoginResponse = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.getUserToken.userToken + " was successfully logged in.", data);
};

gpii.tests.flowManager.getUserToken.testLogoutResponse = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManager.getUserToken.userToken + " was successfully logged out.", data);
};

gpii.tests.flowManager.getUserToken.testInvalidgetUserTokenRequest = function (data) {
    data = JSON.parse(data);
    jqUnit.assertDeepEq("Expecting error messages as reply to http request", {
        isError: true,
        message: "No user currently logged in to the system"
    }, data);
};

gpii.tests.flowManager.getUserToken.testGetUserTokenRequest = function (data) {
    data = JSON.parse(data);
    jqUnit.assertEquals("Expecting token returned in payload", data, gpii.tests.flowManager.getUserToken.userToken);
};

var testDefs = [{
    name: "Flow Manager getUserToken tests.",
    expect: 5,
    config: {
        configName: "getUserToken",
        configPath: configPath
    },
    components: {
        getUserTokenRequest: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    port: "{configuration}.options.port",
                    path: "/userToken"
                }
            }
        },
        login: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    port: "{configuration}.options.port",
                    path: "/user/%userToken/login"
                },
                termMap: {
                    userToken: gpii.tests.flowManager.getUserToken.userToken
                }
            }
        },
        logout: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    port: "{configuration}.options.port",
                    path: "/user/%userToken/logout"
                },
                termMap: {
                    userToken: gpii.tests.flowManager.getUserToken.userToken
                }
            }
        }
    },
    sequence: [{
        funcName: "fluid.pushSoftFailure",
        args: kettle.utils.failureHandler
    }, {
        func: "{getUserTokenRequest}.send"
    }, {
        event: "{getUserTokenRequest}.events.onComplete",
        listener: "gpii.tests.flowManager.getUserToken.testInvalidgetUserTokenRequest"
    }, {
        func: "{login}.send"
    }, {
        event: "{login}.events.onComplete",
        listener: "gpii.tests.flowManager.getUserToken.testLoginResponse"
    }, {
        func: "{getUserTokenRequest}.send"
    }, {
        event: "{getUserTokenRequest}.events.onComplete",
        listener: "gpii.tests.flowManager.getUserToken.testGetUserTokenRequest"
    }, {
        func: "{logout}.send"
    }, {
        event: "{logout}.events.onComplete",
        listener: "gpii.tests.flowManager.getUserToken.testLogoutResponse"
    }, {
        func: "{getUserTokenRequest}.send"
    }, {
        event: "{getUserTokenRequest}.events.onComplete",
        listener: "gpii.tests.flowManager.getUserToken.testInvalidgetUserTokenRequest"
    }, {
        funcName: "fluid.pushSoftFailure",
        args: -1
    }]
}];

module.exports = kettle.test.bootstrapServer(testDefs);
