// flowManager/test/PCPInterfaceTests.js
/**
 * GPII PCP Interface Tests
 *
 * Copyright 2015 Raising the Floor - International
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

"use strict";

var fluid = require("infusion"),
    path = require("path"),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "../configs"),
    kettle = require("kettle"),
    gpii = fluid.registerNamespace("gpii");

require("universal");

kettle.loadTestingSupport();

fluid.registerNamespace("gpii.tests.flowManager.PCPInterface");

gpii.tests.flowManager.PCPInterface.checkConnectionRequest = function (data, client) {
    jqUnit.assertEquals("Connection succeeded", "Client has successfully established connection to the PCP Channel", data);

    // connect socket events to the client events so we can test it
    client.socket.on("login", client.events.onLogin.fire);
    client.socket.on("logout", client.events.onLogout.fire);
    client.socket.on("connectionSucceeded", client.events.connectionSucceeded.fire);
    // client.socket.on("message", client.events.onMessage.fire);
};

fluid.defaults("gpii.tests.flowManager.PCPInterface.basicClient", {
    gradeNames: ["kettle.test.request.io", "autoInit"],
    path: "/pcpChannel",
    port: 8081,
    listenOnInit: true,
    events: {
        onLogin: null,
        onLogout: null,
        onMessage: null,
        connectionSucceeded: null
    },
    listeners: {
        onLogin: {
            funcName: "gpii.tests.flowManager.PCPInterface.onLogin",
            args: [ "{arguments}.0" ]
        }
    }
});

fluid.defaults("gpii.tests.flowManager.pcpInterface.testCaseHolder", {
    gradeNames: ["kettle.test.testCaseHolder", "autoInit"],
    distributeOptions: {
        record: {
            funcName: "gpii.tests.flowManager.pcpInterface.receiveLifecycleManager",
            args: ["{testCaseHolder}", "{arguments}.0"]
        },
        target: "{that flowManager}.options.listeners.onCreate"
    }
});


gpii.tests.flowManager.pcpInterface.receiveLifecycleManager = function (testCaseHolder, flowManager) {
    testCaseHolder.flowManager = flowManager;
};

gpii.tests.flowManager.PCPInterface.checkConnectionSucceeded = function (data, expected) {
    jqUnit.assertDeepEq("Checking data received on connection success: ", expected, data);
};

gpii.tests.flowManager.PCPInterface.onLogin = function (data) {
    jqUnit.assertDeepEq("Testing data on login event", {
        "settings": {
            "http://registry.gpii.net/common/highContrastEnabled": false
        },
        "userToken": "someUser"
    }, data);
};

gpii.tests.flowManager.PCPInterface.onLogout = function () {
    jqUnit.assertTrue("Logout event received on client side of socket", true);
};

gpii.tests.flowManager.PCPInterface.onMessage = function (data, expected) {
    jqUnit.assertDeepEq("Message succusfully received by socket client", expected, data);
};

var testDefs = [
    {
        name: "Flow Manager PCPInterface simple messagetests",
        expect: 2,
        config: {
            configName: "development",
            configPath: configPath
        },
        gradeNames: "gpii.tests.flowManager.pcpInterface.testCaseHolder",
        components: {
            client: {
                type: "gpii.tests.flowManager.PCPInterface.basicClient"
            }
        },
        sequence: [
            {
                func: "{client}.send"
            }, {
                event: "{client}.events.onComplete",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionRequest",
                args: [ "{arguments}.0", "{client}" ]
            }, {
                // Test simpe messages gets passed to socket client
                func: "{flowManager}.pcpInterface.sendUserMessage",
                args: [ "Howdy user! This is a message to you" ]
            }, {
                event: "{client}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "Howdy user! This is a message to you"
                    }
                ]
            }
        ]
    }, {
        name: "Flow Manager PCPInterface with single client login, logout and standard message tests",
        expect: 7,
        config: {
            configName: "development",
            configPath: configPath
        },
        gradeNames: "gpii.tests.flowManager.pcpInterface.testCaseHolder",
        components: {
            client: {
                type: "gpii.tests.flowManager.PCPInterface.basicClient"
            }
        },
        sequence: [
            {
                func: "{client}.send"
            }, {
                event: "{client}.events.onComplete",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionRequest",
                args: [ "{arguments}.0", "{client}" ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client}.events.connectionSucceeded",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionSucceeded",
                args: [ "{arguments}.0", {} ]
            }, {
                // Test regular login notification
                func: "{flowManager}.pcpInterface.notifyUserLogonStateChange",
                args: [ "login", "someUser" ]
            }, {
                event: "{client}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "The token someUser was logged in."
                    }
                ]
            }, {
                // Test simpe messages gets passed to socket client when logged in
                func: "{flowManager}.pcpInterface.sendUserMessage",
                args: [ "Howdy user! This is a message to you" ]
            }, {
                event: "{client}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "Howdy user! This is a message to you"
                    }
                ]
            }, {
                // Test logout notification gets passed to socket client
                func: "{flowManager}.pcpInterface.notifyUserLogonStateChange",
                args: [ "logout", "someUser" ]
            }, {
                event: "{client}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "The token someUser was logged out."
                    }
                ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client}.events.onLogout",
                listener: "gpii.tests.flowManager.PCPInterface.onLogout"
            }
        ]
    }, {
        name: "Flow Manager PCPInterface with single client checking connection succeeded after login",
        expect: 4,
        config: {
            configName: "development",
            configPath: configPath
        },
        gradeNames: "gpii.tests.flowManager.pcpInterface.testCaseHolder",
        components: {
            client: {
                type: "gpii.tests.flowManager.PCPInterface.basicClient"
            }
        },
        sequence: [
            {
                // First login
                func: "{flowManager}.pcpInterface.notifyUserLogonStateChange",
                args: [ "login", "someUser" ]
            }, {  // connect client and check that we get info about logged in user:
                func: "{client}.send"
            }, {
                event: "{client}.events.onComplete",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionRequest",
                args: [ "{arguments}.0", "{client}" ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client}.events.connectionSucceeded",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionSucceeded",
                args: [ "{arguments}.0", {
                    "settings": {
                        "http://registry.gpii.net/common/highContrastEnabled": false
                    },
                    "userToken": "someUser"
                }]
            }, {
                // Test logout notification gets passed to socket client
                func: "{flowManager}.pcpInterface.notifyUserLogonStateChange",
                args: [ "logout", "someUser" ]
            }, {
                event: "{client}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "The token someUser was logged out."
                    }
                ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client}.events.onLogout",
                listener: "gpii.tests.flowManager.PCPInterface.onLogout"
            }
        ]
    }, {
        name: "Flow Manager PCPInterface with multiple clients",
        expect: 8,
        config: {
            configName: "development",
            configPath: configPath
        },
        gradeNames: "gpii.tests.flowManager.pcpInterface.testCaseHolder",
        components: {
            client1: {
                type: "gpii.tests.flowManager.PCPInterface.basicClient"
            },
            client2: {
                type: "gpii.tests.flowManager.PCPInterface.basicClient"
            }
        },
        sequence: [
            { // connect first client
                func: "{client1}.send"
            }, {
                event: "{client1}.events.onComplete",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionRequest",
                args: [ "{arguments}.0", "{client1}" ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client1}.events.connectionSucceeded",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionSucceeded",
                args: [ "{arguments}.0", {} ]
            }, { // log user in
                func: "{flowManager}.pcpInterface.notifyUserLogonStateChange",
                args: [ "login", "someUser" ]
            }, {
                event: "{client1}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "The token someUser was logged in."
                    }
                ]
            }, { // connect second client
                func: "{client2}.send"
            }, {
                event: "{client2}.events.onComplete",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionRequest",
                args: [ "{arguments}.0", "{client2}" ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client2}.events.connectionSucceeded",
                listener: "gpii.tests.flowManager.PCPInterface.checkConnectionSucceeded",
                args: [ "{arguments}.0", {
                    "settings": {
                        "http://registry.gpii.net/common/highContrastEnabled": false
                    },
                    "userToken": "someUser"
                }]
            }, { // Send message (which should be retrieved by both clients)
                func: "{flowManager}.pcpInterface.sendUserMessage",
                args: [ "Howdy user! This is a message to you" ]
            }, {
                event: "{client2}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "Howdy user! This is a message to you"
                    }
                ]
            }, {
                funcName: "fluid.identity"
            }, {
                event: "{client1}.events.onMessage",
                listener: "gpii.tests.flowManager.PCPInterface.onMessage",
                args: [
                    "{arguments}.0", {
                        type: "infoMessage",
                        message: "Howdy user! This is a message to you"
                    }
                ]
            }, {
                // Test logout notification gets passed to socket client
                func: "{flowManager}.pcpInterface.notifyUserLogonStateChange",
                args: [ "logout", "someUser" ]
            }
        ]
    }
];

module.exports = kettle.test.bootstrapServer(testDefs);
