// flowManager/test/SaveTests.js
/**
 * GPII Flow Manager Save Tests
 *
 * Copyright 2013 OCAD University
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

/*global require, __dirname*/

var fluid = require("infusion"),
    path = require("path"),
    flowManager = fluid.require("flowManager", require),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "./configs"),
    kettle = fluid.registerNamespace("kettle"),
    fs = require("fs"),
    statTime = new Date();

fluid.require("kettle/test/utils/js/KettleTestUtils", require);

kettle.tests.testToken1 = "testToken1";
kettle.tests.prefsDir = path.resolve(__dirname,
    "../../../../testData/preferences/");
kettle.tests.testToken1Set = {
    test: "test"
};

kettle.tests.filesToDelete = [];

var cleanUpTmpFiles = function () {
    fluid.each(kettle.tests.filesToDelete, function (filePath) {
        fs.unlinkSync(filePath);
    });
};

kettle.tests.testSaveResponse = function (data) {
    var data = JSON.parse(data);
    var filePath = path.resolve(kettle.tests.prefsDir, data.token) + ".json";
    kettle.tests.filesToDelete.push(filePath);
    jqUnit.assertDeepEq("Response is correct", kettle.tests.testToken1Set,
        data.preferences);
    var savedPrefs = require(filePath);
    jqUnit.assertDeepEq("Saved preferences are correct",
        kettle.tests.testToken1Set, savedPrefs);
};

function jqUnitCondition (args, activity) {
    if (QUnit.config.current) {
        QUnit.ok(false, "Assertion failure (see console.log for expanded message): ".concat(args));
    }
    fluid.builtinFail(false, args, activity);
}

function utilsCondition (args, activity) {
    var messages = args.concat(activity),
        request = fluid.expand("{request}", {
            fetcher: fluid.makeEnvironmentFetcher()
        });
    if (!request) {
        messages = ["ASSERTION FAILED: "].concat(messages);
        console.log.apply(null, messages);
        throw new Error(args[0]);
    }
    request.events.onError.fire({
        isError: true,
        message: messages.join("")
    });
}

kettle.tests.replaceDataSource = function (flowManager) {
    if (kettle.tests.realDataSource) {
        fluid.pushSoftFailure(-1);
        fluid.pushSoftFailure(jqUnitCondition);
        flowManager.preferencesDataSource = kettle.tests.realDataSource;
    } else {
        fluid.pushSoftFailure(-1);
        fluid.pushSoftFailure(utilsCondition);
        kettle.tests.realDataSource = flowManager.preferencesDataSource;
        flowManager.preferencesDataSource = {
            set: function () {
                fluid.fail("Failed data source as expected.");
            }
        };
    }
};

kettle.tests.testBadResponse = function (data) {
    data = JSON.parse(data);
    jqUnit.assertTrue("Received error as expected", data.isError);
    jqUnit.assertTrue("Message is correct",
        data.message.indexOf("Failed data source as expected.") > -1);
};

var testDefs = [{
    name: "Flow Manager save tests.",
    expect: 6,
    config: {
        nodeEnv: "save",
        configPath: configPath
    },
    listeners: {
        onDestroy: cleanUpTmpFiles
    },
    components: {
        saveRequest: {
            type: "kettle.tests.request.http",
            options: {
                requestOptions: {
                    path: "/save/%token",
                    method: "POST"

                },
                termMap: {
                    token: kettle.tests.testToken1
                }
            }
        },
        saveRequestNoToken: {
            type: "kettle.tests.request.http",
            options: {
                requestOptions: {
                    path: "/save",
                    method: "POST"
                }
            }
        }
    },
    sequence: [{
        func: "{saveRequest}.send",
        args: kettle.tests.testToken1Set
    }, {
        event: "{saveRequest}.events.onComplete",
        listener: "kettle.tests.testSaveResponse"
    }, {
        func: "kettle.tests.replaceDataSource",
        args: "{tests}.configuration.server.flowManager"
    }, {
        func: "{saveRequestNoToken}.send"
    }, {
        event: "{saveRequestNoToken}.events.onComplete",
        listener: "kettle.tests.testBadResponse"
    }, {
        func: "kettle.tests.replaceDataSource",
        args: "{tests}.configuration.server.flowManager"
    }, {
        func: "{saveRequestNoToken}.send",
        args: kettle.tests.testToken1Set
    }, {
        event: "{saveRequestNoToken}.events.onComplete",
        listener: "kettle.tests.testSaveResponse"
    }]
}];

module.exports = kettle.tests.bootstrap(testDefs);
