// flowManager/test/UpdateTests.js
/**
 * GPII Flow Manager Update Tests
 *
 * Copyright 2013 OCAD University
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

/*global __dirname */

"use strict";

var fluid = require("infusion"),
    path = require("path"),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "./configs"),
    kettle = fluid.registerNamespace("kettle"),
    gpii = fluid.registerNamespace("gpii"),
    $ = fluid.registerNamespace("jQuery");

fluid.require("kettle/test/utils/js/KettleTestUtils", require);
fluid.require("ontologyHandler", require);

fluid.registerNamespace("gpii.tests.flowManagerUpdate");

gpii.tests.flowManagerUpdate.token = "testUser1";

gpii.tests.flowManagerUpdate.tokenDefaultPrefs = require(path.resolve(__dirname,
    "../../../../testData/preferences/testUser1.json"));

gpii.tests.flowManagerUpdate.tokenPrefs = $.extend(true, {}, gpii.tests.flowManagerUpdate.tokenDefaultPrefs, {
    "http://registry.gpii.net/common/setting1": [{ value: 14 }]
});

gpii.tests.flowManagerUpdate.tokenPrefsUpdate = {
    success: true
};

gpii.tests.flowManagerUpdate.tokenPrefsUpdated = $.extend(true, {}, gpii.tests.flowManagerUpdate.tokenDefaultPrefs, {
    "http://registry.gpii.net/common/setting1": [{ value: 16 }]
});

gpii.tests.flowManagerUpdate.testLoginResponse = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManagerUpdate.token + " was successfully logged in.", data);
};

gpii.tests.flowManagerUpdate.testLogoutResponse = function (data) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        gpii.tests.flowManagerUpdate.token + " was successfully logged out.", data);
};

gpii.tests.flowManagerUpdate.testInvalidUpdateRequest = function () {
    jqUnit.assertTrue("Authorization failed as expected", true);
};

function testUpdateRequest(expected, data) {
    jqUnit.assertDeepEq("Response is correct", expected, data);
}

gpii.tests.flowManagerUpdate.testFirstUpdateRequest = function (data) {
    testUpdateRequest(gpii.tests.flowManagerUpdate.tokenPrefsUpdate, data);
};

gpii.tests.flowManagerUpdate.testSecondUpdateRequest = function (data) {
    testUpdateRequest(gpii.tests.flowManagerUpdate.tokenPrefsUpdate, data);
};

gpii.tests.flowManagerUpdate.clearActiveSessions = function (activeSessions) {
    delete activeSessions.testUser1;
    fluid.pushSoftFailure(function testCondition() {
        fluid.builtinFail(false, ["This failure is expected."]);
    });
};

gpii.tests.flowManagerUpdate.testUpdateRequestFailure = function (data) {
    jqUnit.assertTrue("Received error as expected", data.isError);
    jqUnit.assertTrue("Message is correct",
        data.message.indexOf("none was active") > -1);
    fluid.pushSoftFailure(-1);
};

var testDefs = [{
    name: "Flow Manager update tests.",
    expect: 8,
    config: {
        configName: "update",
        configPath: configPath
    },
    components: {
        invalidUpdateRequest: {
            type: "kettle.test.request.io",
            options: {
                requestOptions: {
                    path: "/update"
                }
            }
        },
        login: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    path: "/user/%token/login"
                },
                termMap: {
                    token: gpii.tests.flowManagerUpdate.token
                }
            }
        },
        logout: {
            type: "kettle.test.request.http",
            options: {
                requestOptions: {
                    path: "/user/%token/logout"
                },
                termMap: {
                    token: gpii.tests.flowManagerUpdate.token
                }
            }
        },
        updateRequest: {
            type: "kettle.test.request.io",
            options: {
                requestOptions: {
                    path: "/update"
                }
            }
        }
    },
    sequence: [{
        func: "{invalidUpdateRequest}.send",
        args: "You shall not pass!!"
    }, {
        event: "{invalidUpdateRequest}.events.onError",
        listener: "gpii.tests.flowManagerUpdate.testInvalidUpdateRequest"
    }, {
        func: "{login}.send"
    }, {
        event: "{login}.events.onComplete",
        listener: "gpii.tests.flowManagerUpdate.testLoginResponse"
    }, {
        func: "{updateRequest}.send",
        args: gpii.tests.flowManagerUpdate.tokenPrefs
    }, {
        event: "{updateRequest}.events.onComplete",
        listener: "gpii.tests.flowManagerUpdate.testFirstUpdateRequest"
    }, {
        func: "{updateRequest}.send",
        args: gpii.tests.flowManagerUpdate.tokenPrefsUpdated
    }, {
        event: "{updateRequest}.events.onComplete",
        listener: "gpii.tests.flowManagerUpdate.testSecondUpdateRequest"
    }, {
        func: "{logout}.send"
    }, {
        event: "{logout}.events.onComplete",
        listener: "gpii.tests.flowManagerUpdate.testLogoutResponse"
    }, {
        func: "{login}.send"
    }, {
        event: "{login}.events.onComplete",
        listener: "gpii.tests.flowManagerUpdate.testLoginResponse"
    }, {
        func: "gpii.tests.flowManagerUpdate.clearActiveSessions",
        args: "{tests}.configuration.server.lifecycleManager.activeSessions"
    }, {
        func: "{updateRequest}.send",
        args: gpii.tests.flowManagerUpdate.tokenPrefs
    }, {
        event: "{updateRequest}.events.onError",
        listener: "gpii.tests.flowManagerUpdate.testUpdateRequestFailure"
    }]
}];

module.exports = kettle.test.bootstrapServer(testDefs);
