// flowManager/test/UpdateTests.js
/**
 * GPII Flow Manager Update Tests
 *
 * Copyright 2013 OCAD University
 *
 * Licensed under the New BSD license. You may not use this file except in
 * compliance with this License.
 *
 * You may obtain a copy of the License at
 * https://github.com/GPII/kettle/LICENSE.txt
 */

/*global require, __dirname*/

var fluid = require("infusion"),
    path = require("path"),
    flowManager = fluid.require("flowManager", require),
    jqUnit = fluid.require("jqUnit"),
    configPath = path.resolve(__dirname, "./configs"),
    kettle = fluid.registerNamespace("kettle"),
    fs = require("fs"),
    $ = fluid.registerNamespace("jQuery");

fluid.require("kettle/test/utils/js/KettleTestUtils", require);

var update = fluid.registerNamespace("kettle.tests.flowManagerUpdate");

update.token = "testUser1";

update.tokenDefaultPrefs = require(path.resolve(__dirname,
    "../../../../testData/preferences/testUser1.json"));

update.tokenPrefs = $.extend(true, {}, update.tokenDefaultPrefs, {
    "http://registry.gpii.org/common/setting1": [{ value: 14 }]
});

update.tokenPrefsUpdate = {
    success: true
};

update.tokenPrefsUpdated = $.extend(true, {}, update.tokenDefaultPrefs, {
    "http://registry.gpii.org/common/setting1": [{ value: 16 }]
});

update.testLoginResponse = function (data, headers, cookies, signedCookies) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        update.token + " was successfully logged in.", data);
    jqUnit.assertValue("Cookie is set", headers["set-cookie"]);
    jqUnit.assertTrue("kettle session cookie is set", !!signedCookies["kettle.sid"]);
};

update.testLogoutResponse = function (data, headers, cookies) {
    jqUnit.assertEquals("Response is correct", "User with token " +
        update.token + " was successfully logged out.", data);
    jqUnit.assertEquals("kettle session cookie is unset", "", cookies["kettle.sid"]);
};

update.testInvalidUpdateRequest = function (reason) {
    jqUnit.assertTrue("Authorization failed as expected", true);
};

update.testFirstUpdateRequest = function (data) {
    testUpdateRequest(update.tokenPrefsUpdate, data)
};

update.testSecondUpdateRequest = function (data) {
    testUpdateRequest(update.tokenPrefsUpdate, data)
};

update.clearActiveSessions = function (activeSessions) {
    delete activeSessions.testUser1;
    fluid.pushSoftFailure(function testCondition() {
        fluid.builtinFail(false, ["This failure is expected."]);
    });
};

update.testUpdateRequestFailure = function (data) {
    jqUnit.assertTrue("Received error as expected", data.isError);
    jqUnit.assertTrue("Message is correct",
        data.message.indexOf("This failure is expected.") > -1);
    fluid.pushSoftFailure(-1);
};

function testUpdateRequest(expected, data) {
    jqUnit.assertDeepEq("Response is correct", expected, data);
}

var testDefs = [{
    name: "Flow Manager update tests.",
    expect: 13,
    config: {
        nodeEnv: "update",
        configPath: configPath
    },
    components: {
        invalidUpdateRequest: {
            type: "kettle.tests.request.io",
            options: {
                requestOptions: {
                    path: "/update"
                }
            }
        },
        login: {
            type: "kettle.tests.request.http",
            options: {
                requestOptions: {
                    path: "/user/%token/login"
                },
                termMap: {
                    token: update.token
                }
            }
        },
        logout: {
            type: "kettle.tests.request.http",
            options: {
                requestOptions: {
                    path: "/user/%token/logout"
                },
                termMap: {
                    token: update.token
                }
            }
        },
        updateRequest: {
            type: "kettle.tests.request.io",
            options: {
                requestOptions: {
                    path: "/update"
                }
            }
        }
    },
    sequence: [{
        func: "{invalidUpdateRequest}.send",
        args: "You shall not pass!!"
    }, {
        event: "{invalidUpdateRequest}.events.onError",
        listener: "kettle.tests.flowManagerUpdate.testInvalidUpdateRequest"
    }, {
        func: "{login}.send"
    }, {
        event: "{login}.events.onComplete",
        listener: "kettle.tests.flowManagerUpdate.testLoginResponse"
    }, {
        func: "{updateRequest}.send",
        args: update.tokenPrefs
    }, {
        event: "{updateRequest}.events.onComplete",
        listener: "kettle.tests.flowManagerUpdate.testFirstUpdateRequest"
    }, {
        func: "{updateRequest}.send",
        args: update.tokenPrefsUpdated
    }, {
        event: "{updateRequest}.events.onComplete",
        listener: "kettle.tests.flowManagerUpdate.testSecondUpdateRequest"
    }, {
        func: "{logout}.send"
    }, {
        event: "{logout}.events.onComplete",
        listener: "kettle.tests.flowManagerUpdate.testLogoutResponse"
    }, {
        func: "{login}.send"
    }, {
        event: "{login}.events.onComplete",
        listener: "kettle.tests.flowManagerUpdate.testLoginResponse"
    }, {
        func: "kettle.tests.flowManagerUpdate.clearActiveSessions",
        args: "{tests}.configuration.server.lifecycleManager.activeSessions"
    }, {
        func: "{updateRequest}.send",
        args: update.tokenPrefs
    }, {
        event: "{updateRequest}.events.onComplete",
        listener: "kettle.tests.flowManagerUpdate.testUpdateRequestFailure"
    }]
}];

module.exports = kettle.tests.bootstrap(testDefs);
