// rawPreferencesServer/test/RawPreferencesTest.js
/**
GPII Raw Preferences Server Test

Copyright 2014 Raising the Floor - International

Licensed under the New BSD license. You may not use this file except in
compliance with this License.

The research leading to these results has received funding from the European Union's
Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.

You may obtain a copy of the License at
https://github.com/GPII/universal/blob/master/LICENSE.txt
*/

"use strict";

var fluid = require("infusion"),
    path = require("path"),
    jqUnit = fluid.require("jqUnit"),
    kettle = fluid.registerNamespace("kettle"),
    gpii = fluid.registerNamespace("gpii"),
    fs = require("fs");

require("universal");

gpii.loadTestingSupport();

fluid.require("rawPreferencesServer", require);

fluid.registerNamespace("gpii.rawPreferencesServer.tests");
gpii.rawPreferencesServer.tests.testDefs = [];

/* ---------------------- TESTING GET FUNCTIONALITY --------------------------- */
fluid.registerNamespace("gpii.rawPreferencesServer.tests.get");

gpii.rawPreferencesServer.tests.get.successfulRequestAssert = function (userToken) {
    return function (data) {
        data = JSON.parse(data);

        // read expected preferences directly from preferences file
        var rawPrefs = fs.readFileSync(__dirname + "/data/" + userToken + ".json");
        var expected = {
            userToken: userToken,
            preferences: JSON.parse(rawPrefs)
        };

        jqUnit.assertDeepEq("Response is correct", expected, data);
    };
};

gpii.rawPreferencesServer.tests.get.failingRequestAssert = function (data) {
    data = JSON.parse(data);
    // If a user token doesn't exist, expect the preferences part to be undefined
    jqUnit.assertTrue("Checking that we're getting an error for nonexisting user token", data.isError);
    jqUnit.assertTrue("Checking that we're getting a 404 error code for nonexisting user token", data.statusCode === 404);
};

gpii.rawPreferencesServer.tests.get.buildTestDef = function (fixture) {
    var sequence = [{
        func: "{getRequest}.send"
    }];
    sequence.push(fixture.onCompleteSequence);

    return {
        name: fixture.name,
        expect: fixture.expect,
        config: {
            configName: "testSetup",
            configPath: "%universal/gpii/node_modules/rawPreferencesServer/configs"
        },
        components: {
            getRequest: {
                type: "kettle.test.request.http",
                options: {
                    path: "/rawPreferences/%userToken",
                    port: 8078,
                    termMap: {
                        userToken: fixture.userToken
                    }
                }
            }
        },
        sequence: sequence
    };
};

gpii.rawPreferencesServer.tests.get.fixtures = [
    {
        name: "Raw Preferences Server - basic GET test",
        expect: 1,
        userToken: "np_flatonly",
        onCompleteSequence: {
            event: "{getRequest}.events.onComplete",
            listenerMaker: "gpii.rawPreferencesServer.tests.get.successfulRequestAssert",
            makerArgs: [ "np_flatonly" ]
        }
    }, {
        name: "Raw Preferences Server - empty NP set",
        expect: 1,
        userToken: "np_empty",
        onCompleteSequence: {
            event: "{getRequest}.events.onComplete",
            listenerMaker: "gpii.rawPreferencesServer.tests.get.successfulRequestAssert",
            makerArgs: [ "np_empty" ]
        }
    }, {
        name: "Raw Preferences Server - failing GET test (non-existing user token)",
        expect: 2,
        userToken: "bogus_np",
        onCompleteSequence: {
            event: "{getRequest}.events.onComplete",
            listener: "gpii.rawPreferencesServer.tests.get.failingRequestAssert"
        }
    }
];

fluid.each(gpii.rawPreferencesServer.tests.get.fixtures, function (fixture) {
    gpii.rawPreferencesServer.tests.testDefs.push(gpii.rawPreferencesServer.tests.get.buildTestDef(fixture));
});

/* ---------------------- TESTING POST FUNCTIONALITY --------------------------- */
fluid.registerNamespace("gpii.rawPreferencesServer.tests.post");

gpii.rawPreferencesServer.tests.post.assertNewNPResponse = function (data) {
    data = JSON.parse(data);
    jqUnit.assertTrue("Structure of returned payload should contain user token and preferences",
        data.preferences !== undefined && data.userToken !== undefined);
    // saved prefs set should contain the preferences
    var fsPrefs = fs.readFileSync(__dirname + "/data/" + data.userToken + ".json");
    jqUnit.assertDeepEq("Expecting preferences to be those posted response",
        JSON.parse(fsPrefs), data.preferences);

    // clean up the created NP set
    fs.unlinkSync(__dirname + "/data/" + data.userToken + ".json");
};

gpii.rawPreferencesServer.tests.post.buildTestDef = function (fixture) {
    return {
        name: fixture.name,
        expect: 2,
        config: {
            configName: "testSetup",
            configPath: "%universal/gpii/node_modules/rawPreferencesServer/configs"
        },
        components: {
            postRequest: {
                type: "kettle.test.request.http",
                options: {
                    path: "/rawPreferences",
                    method: "POST",
                    port: 8078
                }
            }
        },
        sequence: [{
            func: "{postRequest}.send",
            args: fixture.postBody
        }, {
            event: "{postRequest}.events.onComplete",
            listener: "gpii.rawPreferencesServer.tests.post.assertNewNPResponse"
        }]
    };
};

gpii.rawPreferencesServer.tests.post.fixtures = [
    {
        name: "POST: With no user token specified",
        postBody: {
            "flat": {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/common/onScreenKeyboardEnabled": [{ "value": true }],
                            "http://registry.gpii.net/common/initDelay": [{ "value": 0.120 }],
                            "http://registry.gpii.net/common/cursorSpeed": [{ "value": 0.850 }],
                            "http://registry.gpii.net/common/cursorAcceleration": [{ "value": 0.800 }],
                            "http://registry.gpii.net/common/mouseEmulationEnabled": [{ "value": true }],
                            "http://registry.gpii.net/common/unknown": [{ "value": true }]
                        }
                    }
                },
                "metadata": []
            }
        }
    }
];

fluid.each(gpii.rawPreferencesServer.tests.post.fixtures, function (fixture) {
    gpii.rawPreferencesServer.tests.testDefs.push(gpii.rawPreferencesServer.tests.post.buildTestDef(fixture));
});


/* ---------------------- TESTING PUT FUNCTIONALITY --------------------------- */
fluid.registerNamespace("gpii.rawPreferencesServer.tests.put");

gpii.rawPreferencesServer.tests.put.assertResponse = function (userToken) {
    return function (data) {
        data = JSON.parse(data);
        var fsPrefs = fs.readFileSync(__dirname + "/data/" + userToken + ".json");

        jqUnit.assertDeepEq("Expecting return payload to contain prefs and user token",
             { userToken: userToken, preferences: JSON.parse(fsPrefs) }, data);
    };
};

gpii.rawPreferencesServer.tests.put.copyNPSet = function (from, to) {
    var fsPrefs = fs.readFileSync(__dirname + "/data/" + from + ".json");
    fs.writeFileSync(__dirname + "/data/" + to + ".json", fsPrefs);
};

gpii.rawPreferencesServer.tests.put.deleteNPSet = function (userToken) {
    fs.unlinkSync(__dirname + "/data/" + userToken + ".json");
};

gpii.rawPreferencesServer.tests.put.buildTestDef = function (fixture) {
    var testDef = {
        name: fixture.name,
        expect: 1,
        config: {
            configName: "testSetup",
            configPath: "%universal/gpii/node_modules/rawPreferencesServer/configs"
        },
        components: {
            postRequest: {
                type: "kettle.test.request.http",
                options: {
                    path: "/rawPreferences/%userToken",
                    method: "PUT",
                    port: 8078,
                    termMap: {
                        userToken: fixture.userToken
                    }
                }
            }
        },
        sequence: [{
            func: "{postRequest}.send",
            args: fixture.putBody
        }, {
            event: "{postRequest}.events.onComplete",
            listenerMaker: "gpii.rawPreferencesServer.tests.put.assertResponse",
            makerArgs: [ fixture.userToken ]
        }, {
            func: "gpii.rawPreferencesServer.tests.put.deleteNPSet",
            args: [ fixture.userToken ]
        }]
    };

    if (fixture.preSequence) {
        testDef.sequence.unshift(fixture.preSequence);
    }
    return testDef;

};

gpii.rawPreferencesServer.tests.put.fixtures = [
    {
        name: "PUT: Using user token for new preferences set",
        userToken: "brand_new",
        putBody: {
            "flat": {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/common/onScreenKeyboardEnabled": [{ "value": true }],
                            "http://registry.gpii.net/common/mouseEmulationEnabled": [{ "value": true }],
                            "http://registry.gpii.net/common/unknown": [{ "value": true }]
                        }
                    }
                },
                "metadata": []
            }
        }
    }, {
        name: "PUT: Updating existing preferences set",
        userToken: "overwriter",
        putBody: {
            "flat": {
                "contexts": {
                    "gpii-default": {
                        "name": "Default preferences",
                        "preferences": {
                            "http://registry.gpii.net/common/onScreenKeyboardEnabled": [{ "value": true }],
                            "http://registry.gpii.net/common/initDelay": [{ "value": 0.120 }],
                            "http://registry.gpii.net/common/cursorSpeed": [{ "value": 0.850 }],
                            "http://registry.gpii.net/common/cursorAcceleration": [{ "value": 0.800 }],
                            "http://registry.gpii.net/common/mouseEmulationEnabled": [{ "value": true }],
                            "http://registry.gpii.net/common/unknown": [{ "value": true }]
                        }
                    }
                },
                "metadata": []
            }
        },
        preSequence: {
            func: "gpii.rawPreferencesServer.tests.put.copyNPSet",
            args: [ "np_flatonly", "overwriter" ]
        }
    }
];

fluid.each(gpii.rawPreferencesServer.tests.put.fixtures, function (fixture) {
    gpii.rawPreferencesServer.tests.testDefs.push(gpii.rawPreferencesServer.tests.put.buildTestDef(fixture));
});

module.exports = kettle.test.bootstrapServer(gpii.rawPreferencesServer.tests.testDefs);
